src/r/recordFactories/ProjectFactory.ts

```diff
@@ -1032,6 +1047,201 @@ export class ProjectUtils {
             }
           }
 
+          this.actionDict[ta.props.action as string] ?
+            this.actionDict[ta.props.action as string].push(ruleId) :
+            this.actionDict[ta.props.action as string] = [ruleId]
+
+          switch (ta.props.action) {
+            case RuleAction.open_url:
```

Comment:
you can club a lot of case in a single logical block. Most of these run through similar logic.